metashare/api/jobs.py

```diff
@@ -0,0 +1,196 @@
+
+def create_branches_on_github(*, user, repo_url, project, task, repo_root):
+    """
+    Expects to be called in the context of a local github checkout.
+    """
+    gh = login(token=user.gh_token)
+    owner, repo = extract_owner_and_repo(repo_url)
+    repository = gh.repository(owner, repo)
+
+    # Make project branch:
+    if project.branch_name:
+        project_branch_name = project.branch_name
+    else:
+        prefix = get_cumulus_prefix(
+            repo_root=repo_root,
+            repo_name="Repo Name",  # A placeholder, not used.
+            repo_url=repo_url,
+            repo_owner=owner,
+            repo_branch=repository.default_branch,
+            repo_commit=repository.branch(repository.default_branch).latest_sha(),
+        )
+        project_branch_name = f"{prefix}{slugify(project.name)}"
+        project_branch_name = try_to_make_branch(
+            repository,
+            new_branch=project_branch_name,
+            base_branch=repository.default_branch,
+        )
+        project.branch_name = project_branch_name
+
+    # Make task branch:
+    if task.branch_name:
+        task_branch_name = task.branch_name
+    else:
+        task_branch_name = try_to_make_branch(
+            repository,
+            new_branch=f"{project_branch_name}__{slugify(task.name)}",
+            base_branch=project_branch_name,
+        )
+        task.branch_name = task_branch_name
+
+    # We'll then save these in the orchestrating function, once all
+    # tasks are done.
+
+    return task_branch_name
+
+
+def run_appropriate_flow(
+    scratch_org, *, user, repo_root, repo_url, repo_branch, org_config
+):
+    # TODO: I hope this function is right, but I am not whether it is or
+    # how to meaningfully test it!
+    from .models import SCRATCH_ORG_TYPES
+
+    cases = {SCRATCH_ORG_TYPES.Dev: "dev_org", SCRATCH_ORG_TYPES.QA: "qa_org"}
+
+    gh = login(token=user.gh_token)
+    owner, repo = extract_owner_and_repo(repo_url)
+    repository = gh.repository(owner, repo)
+    cci = MetaShareCCI(
+        repo_root=repo_root,
+        repo_name="Repo Name",  # A placeholder, not used.
+        repo_url=repo_url,
+        repo_owner=owner,
+        repo_branch=repo_branch,
+        repo_commit=repository.branch(repo_branch).latest_sha(),
+    )
+    flow_config = cci.get_flow(cases[scratch_org.org_type])
+    flowrunner.FlowCoordinator(cci.project_config, flow_config).run(org_config)
+
+
+def create_branches_on_github_then_create_scratch_org(
+    *, project, repo_url, scratch_org, task, user
+):
+    with local_github_checkout(user, repo_url) as repo_root:
+        commit_ish = create_branches_on_github(
+            user=user,
+            repo_url=repo_url,
+            project=project,
+            task=task,
+            repo_root=repo_root,
+        )
+        org_config = create_scratch_org(
+            scratch_org=scratch_org, user=user, repo_url=repo_url, commit_ish=commit_ish
+        )
+        run_appropriate_flow(
```

Comment:
@wlonk I think this (and maybe even `create_branches_on_github`) should be within a `with report_errors_on(scratch_org):` block?